### PR TITLE
[TensorPipe] Do not require user to provide worker name-to-rank map

### PR DESCRIPTION
**Pull request by lw**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#38052 [TensorPipe] Do not require user to provide worker name-to-rank map**

The initial version of the TensorPipe agent required the user to specify the full map between workers' names and their ids, on each worker. However it's enough for each worker to just specify their name and id, as these can then be exchanged using the store.

Addresses #37784, although I think we can go further and use the store to also automatically assign ranks to workers, so that the user only needs to specify a name.

Differential Revision: [D21463833](https://our.internmc.facebook.com/intern/diff/D21463833/)

### CODE REVIEW

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp [hunk 1]
```diff
@@ -19,10 +19,36 @@ namespace rpc {
...
 constexpr long kToMilliseconds = 1000;
 
+void TensorPipeAgent::collectNames() {
+  const worker_id_t selfId = workerInfo_.id_;
+  const std::string& selfName = workerInfo_.name_;
+
+  std::vector<uint8_t> selfNameVector(
+      (uint8_t*)selfName.c_str(),
+      (uint8_t*)selfName.c_str() + selfName.length());
+  addressStore_->set("names/" + std::to_string(selfId), selfNameVector);
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/tensorpipe_agent.cpp:29]: nit: There is a convention to use `c10::to_string` in our code base.

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp [hunk 2]
```diff
@@ -19,10 +19,36 @@ namespace rpc {
...
+
+  workerIdToInfo_.emplace(selfId, WorkerInfo(selfName, selfId));
+  workerNameToInfo_.emplace(selfName, WorkerInfo(selfName, selfId));
+  for (worker_id_t workerId = 0; workerId < worldSize_; ++workerId) {
+    if (workerId == selfId) {
+      continue;
+    }
+    std::string workerKey = "names/" + std::to_string(workerId);
+    addressStore_->wait({workerKey});
+    std::vector<uint8_t> workerNameVector = addressStore_->get(workerKey);
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/tensorpipe_agent.cpp:39]: Seems `get()` is implicitly doing `wait()`?

File: torch/csrc/distributed/rpc/tensorpipe_agent.h [hunk 3]
```diff
@@ -32,9 +26,10 @@ struct TensorPipeRpcBackendOptions : public RpcBackendOptions {
 class TensorPipeAgent : public RpcAgent {
  public:
   TensorPipeAgent(
-      worker_id_t selfId,
-      std::string selfName,
       std::shared_ptr<::c10d::Store> addressStore,
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/tensorpipe_agent.h:29]: cc @kiukchung @tierex `TensorPipeAgent` is already taking a c10d::Store

**Review by mrshenli — APPROVED** [review]
Thanks for adding this. Test failures do not look relevant, but if master is getting better early tomorrow. Could you please rebase to make sure we don't break anything?

Another thing is that, test coverage becomes more and more urgent. I am timid to keep adding code with no test coverage. It will be great to see TensorPipe getting connected to (subset of) our existing RPC tests.

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp [hunk 4]
```diff
@@ -19,10 +19,36 @@ namespace rpc {
...
+  }
+}
+
 TensorPipeAgent::TensorPipeAgent(
-    worker_id_t selfId,
-    std::string selfName,
     std::shared_ptr<::c10d::Store> addressStore,
+    std::string selfName,
+    worker_id_t selfId,
+    worker_id_t worldSize,
```
Comment (osalpekar) [inline torch/csrc/distributed/rpc/tensorpipe_agent.cpp:51]: I think we use `int` for `worldSize` instead of `worker_id_t`? Looks like worker_id_t is used only for ID's.

**Review by osalpekar — APPROVED** [review]

**Comment by osalpekar** [comment]
> Thanks for adding this. Test failures do not look relevant, but if master is getting better early tomorrow. Could you please rebase to make sure we don't break anything?
> 
> Another thing is that, test coverage becomes more and more urgent. I am timid to keep adding code with no test coverage. It will be great to see TensorPipe getting connected to (subset of) our existing RPC tests.

@mrshenli I have a diff enabling the RPC, Dist Autograd, and Dist Optimizer tests with the TensorPipe Agent (D21412016). Merging this PR will make the tests run more smoothly in the current framework (no need to set a hacky name-to-rank map).

Also, will these tests be push-blocking? A lot of these tests will fail due to unimplemented functionality, so just want to ensure that PR's don't get reverted due to these failures. Otherwise we can disable the failing tests and enable them as we implement more features.

**Comment by lw** [comment]
Uhm, it looks like there has been a ghexport mishap, let me try again...

**Comment by lw** [comment]
All tests are passing, I'm landing this.

I also feel a bit queasy about the lack of tests for the TensorPipe agent. I have ran manual tests for this diff (they are in the test plan on Phabricator, not sure why they didn't get exported). I'm looking forward to enrolling this agent into the existing tests in some form, so that we can have more automatic signal...

**Comment by mrshenli** [comment]
> Also, will these tests be push-blocking? A lot of these tests will fail due to unimplemented functionality, so just want to ensure that PR's don't get reverted due to these failures. Otherwise we can disable the failing tests and enable them as we implement more features.

Thanks @osalpekar for adding the test. I don't think our RPC tests are push-blocking, but if they consistently fail due to missing features, it's better that we only run the subset that can pass currently? Is it possible to create a black list of tests to skip for TensorPipeAgent, and gradually remove tests from that list?
